Compliance Controller Improvements

AdamPettinger opened the pull request:
This is an update to this pull request to kinetic-devel: [(WIP) New type of compliance controller #393](https://github.com/ros-controls/ros_controllers/pull/393).

Updated to Melodic, added singularity checking, and improved directionally constrained compliance.

Tested on UR5 with ATI Gamma Force/Torque sensor. Videos showing improved directional control:

1. [Compliant only in EEF Forward and EEF roll directions](https://drive.google.com/file/d/1i77SVG8YFoIGYAzKRylypvYj8J8b838f/view?usp=sharing)
2. [Grasping a valve with compliance](https://drive.google.com/file/d/1FAsqDhwRbRmkkmAsjUnJbrf_N2DZzJQ1/view?usp=sharing) and no other user input

7675t committed:
Use a copy of the pointer in update()

agutenkunst committed:
Changes to allow inheritance from JointTrajectoryController.

k-okada committed:
isPublishngCmdVelOut to check getNumPublisheres until timeout

k-okada committed:
joint_trajectory_controller_test.cpp adds waitForNextStates, because we want to wait stateCB after waitForState find new states

k-okada committed:
use short_timeout in jointVelocityFeedForward

k-okada committed:
emptyTopicCancelsActionTraj: make sure that stateCB received the newer topics than when we confirmed with waitForState function

k-okada committed:
reloadController, forget to explictly 'return true'

k-okada committed:
ignoreOldTopicTraj add waitForNextState

k-okada committed:
add gtest_filter argument

bmagyar committed:
Update changelogs

bmagyar committed:
0.13.3

graiola committed:
Merge pull request #335 from PilzDE/feature/inheritable_joint_trajectory_controller

Changes to allow inheritance from JointTrajectoryController.

miguelprada committed:
Return error string when failing to initialize trajectory from message

miguelprada committed:
Remove redundant warning messages

miguelprada committed:
Report errors in updateTrajectoryCommand back though action result error_string

bmagyar committed:
Update changelogs

bmagyar committed:
0.13.4

mathias-luedtke committed:
implemented setErrorString

jschleicher committed:
don't print warning about dropped first point (#366)

Commit:
Use stringstream, std::to_string only in c++11

oka1125 committed:
Use a copy of the pointer in update() to avoid crash by cancelCB()

MoriKen254 committed:
Add ackermann_steering_controller (#356)

mathias-luedtke committed:
travis_wait is not needed anymore

mathias-luedtke committed:
travis yml clean-up

AndyZe committed:
Add a new compliance controller.

AndyZe committed:
Implementing msg architecture for joint limits.

AndyZe committed:
Checking joint limits during publication of compliant commands.

AndyZe committed:
Renaming wrench_to_twist_pub more appropriately.

AndyZe committed:
WIP: Parameterizing the outgoing joint velocity topic.

AndyZe committed:
Minor renaming of files.

AndyZe committed:
Adding what's intended to be a temporary usage guide for compliance.

AndyZe committed:
Better comment re. jacobian_frame_name

AndyZe committed:
Clang formatting, adding a (placeholder) service to toggle compliance.

AndyZe committed:
Merge branch 'kinetic-compliance' of https://github.com/UTNuclearRobotics/ros_controllers into kinetic-compliance

AndyZe committed:
Completed the service call to toggle compliance.

AndyZe committed:
WIP: add damping

AndyZe committed:
WIP: breaking the filter out into a separate cpp file, add empty low-pass derivative class.

AndyZe committed:
Implement damping with a filtered derivative of wrench.

AndyZe committed:
Some simplifications in compliant_control.

AndyZe committed:
Got a definitive answer on the sign of damping.

AndyZe committed:
Correct a note in the yaml file.

AndyZe committed:
Small bug on compliant trajectories, and switching the damping sign back to negative after testing it.

AndyZe committed:
Convert the Readme to markdown.

AndyZe committed:
Reset the action server when compliance is toggled.

AndyZe committed:
Use a pseudoinverse that's more stable near singularities.

AndyZe committed:
Testing - drop orientation.

AndyZe committed:
Debugging on indices.

AndyZe committed:
Clang format.

AndyZe committed:
Done parsing dropped_dof_index, need to add remaining rows.

AndyZe committed:
Short-circuiting some loop iterations.

AndyZe committed:
Tweaks to work when dropping the first dimension.

AndyZe committed:
Selectable dimensions seems to work in all cases.

AndyZe committed:
Starting a service to adjust compliance settings on the fly.

AndyZe committed:
Allow for empty dof_to_drop_ commands.

AndyZe committed:
Resize the cartesian_velocity vector, too.

AndyZe committed:
Cleaning up some deprecated drop_dof code.

AndyZe committed:
Moving code out of header, into cpp.

AndyZe committed:
Separating service files by functionality.

AndyZe committed:
Filling the callbacks for setStiffness and setDamping.

AndyZe committed:
Update controller description.

AndyZe committed:
Better usage of return values on damping/stiffness services.

PaperFanz committed:
remove deprecated packages

PaperFanz committed:
add skid-steer-controller needed for husky_start

AndyZe committed:
Remove skid_steer controller

AdamPettinger committed:
Initial commit adding singularity checking

PaperFanz committed:
Singularity checking debugged and tested. Working as expected

AdamPettinger committed:
Fixed and tested dropping compliance in specific dimensions

AdamPettinger committed:
Disable compliance dimensions message update, testing

AdamPettinger committed:
Merged from melodic-compliance

bmagyar commented:
Thanks for this guys. Would it be possible to squash/rebase this PR to reduce the number of commits (and potential conflicts)?

bmagyar commented on compliance_controller/README.md:
```diff
@@ -0,0 +1,111 @@
+
+* Clone ur_modern_driver (kinetic-devel branch), https://github.com/ros-industrial/ur_modern_driver
+
+* Clone NRG's ros_controllers repo (kinetic-compliance branch), https://github.com/UTNuclearRobotics/ros_controllers
```
Needs an update?

bmagyar commented on compliance_controller/compliance_control_msgs/package.xml:
```diff
@@ -0,0 +1,15 @@
+  <version>0.0.0</version>
+  <description>The compliance_control_msgs package</description>
+
+  <maintainer email="andyz@utexas.edu">AndyZe</maintainer>
```
Missing author

bmagyar commented on compliance_controller/wrench_to_joint_vel_pub/include/wrench_to_joint_vel_pub/compliant_control.h:
```diff
@@ -0,0 +1,163 @@
+ */
+enum Dimension
+{
+  NUM_DIMS = 6  // 3 translational, 3 rotational dimensions
```
How about a single cost instead?

bmagyar commented on compliance_controller/wrench_to_joint_vel_pub/src/wrench_to_joint_vel_pub/wrench_to_joint_vel_pub.cpp:
```diff
@@ -0,0 +1,404 @@
+    compliance_control_msgs::AdjustDamping::Request& req,
+    compliance_control_msgs::AdjustDamping::Response& res)
+{
+  std::vector<double> damping = req.damping.data;
```
Const

bmagyar commented on velocity_controllers/CHANGELOG.rst:
```diff
@@ -21,6 +21,10 @@ Changelog for package velocity_controllers
 -------------------
 
 0.14.0 (2018-04-27)
+0.13.4 (2018-06-26)
```
Don't make changes to this file please

bmagyar commented on traj_or_jog_controller/include/traj_or_jog_controller/traj_or_jog_controller.h:
```diff
@@ -0,0 +1,90 @@
+};
+}
+
+#endif  // header guard
```
On melodic you have C++14 by default, this archaic tech can now be replaced by `#pragma once`

bmagyar commented on ros_controllers/CHANGELOG.rst:
```diff
@@ -19,6 +19,10 @@ Changelog for package ros_controllers
 0.14.0 (2018-04-27)
+0.13.4 (2018-06-26)
+-------------------
+
```
No meddling with changelogs please

bmagyar reviewed (changes requested):
First of all, thanks for contributing!! 

Please ensure that you only change code in the files you are adding. Specifically changelogs shouldn't be touched. There's a new Git modules file added, please get rid of it. Make sure you are listed as author and maintainer in each package and please add me as maintainer as well.

I haven't finished with the review but wanted to post these comments so you have some feedback.

bmagyar commented on compliance_controller/compliance_controller/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+
+install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION})
+
```
We proud ourselves here for our tests. They ensure that continued improvements to the controllers and the ros_control framework keep everything working. Could you please add some rostests that ensure some basic functionality? Look at joint traj controller or diff drive controller for good examples.

AdamPettinger commented:
Thanks for the feedback, I am making the requested changes. In terms of squashing it down, I am struggling to do so. This is my first big PR and unfortunately is a messy one :(. A lot of the early commits are not related to the compliance controller and are causing me problems.

Locally I was able to start with a fresh `ros-controls:melodic-devel` branch, and cherry pick the commits for the compliance controller. Then I rebased to squash it to 1-2 commits and everything looks good (only files changed are the ones we added, builds and runs as expected). At this point I am not sure if it is better to merge this branch with our `UTNuclearRobotics:melodic-devel` branch and force push it to update this PR, or to close the PR and open a new one, or to do something else entirely. Do you have insights here?

As for actual code changes, I will work to incorporate the changes requested here, as well as in the original PR for the kinetic branch

bmagyar commented:
First of all I suggest backing it all up. Avoid force pushing to your main branch, create a separate one for this MR and if you want to merge the end result back once its emerged here, so be it. 
Github allows for retargeting PRs including the source if I remember well... 

You should be able to do an interactive rebase. I suggest to do it in a couple of steps. First, review the commits that show up on this PR and drop all that don't belong to your team. This will eliminate the changelog changes and other artifacts of releasing. Then as a second step see if there are hacky commits (the message sometimes tells you) and drop those too. Now, try squashing the first few commits and move on from here. If you have massive conflicts, you may want to rethink your strategy and see if the commit causing it can be eliminated first.

AdamPettinger commented:
Closed this PR to open a new, cleaner one with proper rebasing, some additional features, and some of the fixes requested here

bmagyar commented:
Thank you!

On Wed, 20 Nov 2019, 18:00 AdamPettinger, <notifications@github.com> wrote:

> Closed this PR to open a new, cleaner one with proper rebasing, some
> additional features, and some of the fixes requested here
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-controls/ros_controllers/pull/434?email_source=notifications&email_token=AA24PYMMIDAC6TNG5LPQQGLQUV3K5A5CNFSM4JG7WEGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEETJ3MI#issuecomment-556178865>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA24PYJUKRT4FCEWIDSD5PLQUV3K5ANCNFSM4JG7WEGA>
> .
>
